File: src/uk/org/rivernile/edinburghbustracker/android/BusStopMapActivity.java
/*
 * Copyright (C) 2012 - 2013 Niall 'Rivernile' Scott
 *
 * This software is provided 'as-is', without any express or implied
 * warranty.  In no event will the authors or contributors be held liable for
 * any damages arising from the use of this software.
 *
 * The aforementioned copyright holder(s) hereby grant you a
 * non-transferrable right to use this software for any purpose (including
 * commercial applications), and to modify it and redistribute it, subject to
 * the following conditions:
 *
 *  1. This notice may not be removed or altered from any file it appears in.
 *
 *  2. Any modifications made to this software, except those defined in
 *     clause 3 of this agreement, must be released under this license, and
 *     the source code of any modifications must be made available on a
 *     publically accessible (and locateable) website, or sent to the
 *     original author of this software.
 *
 *  3. Software modifications that do not alter the functionality of the
 *     software but are simply adaptations to a specific environment are
 *     exempt from clause 2.
 */

package uk.org.rivernile.edinburghbustracker.android;

import android.app.SearchManager;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.ActionBarActivity;
import android.view.MenuItem;
import com.google.android.gms.maps.model.LatLng;
import uk.org.rivernile.android.utils.NavigationUtils;
import uk.org.rivernile.edinburghbustracker.android.fragments.dialogs
        .IndeterminateProgressDialogFragment;
import uk.org.rivernile.edinburghbustracker.android.fragments.dialogs
        .MapTypeChooserDialogFragment;
import uk.org.rivernile.edinburghbustracker.android.fragments.dialogs
        .ServicesChooserDialogFragment;
import uk.org.rivernile.edinburghbustracker.android.fragments.general
        .BusStopMapFragment;

public class BusStopMapActivity extends ActionBarActivity
        implements BusStopMapFragment.Callbacks,
        MapTypeChooserDialogFragment.Callbacks,
        ServicesChooserDialogFragment.Callbacks,
        IndeterminateProgressDialogFragment.Callbacks {
    
    /** The stopCode argument for the Intent. */
    public static final String ARG_STOPCODE = BusStopMapFragment.ARG_STOPCODE;
    /** The latitude argument for the Intent. */
    public static final String ARG_LATITUDE = BusStopMapFragment.ARG_LATITUDE;
    /** The longitude argument for the Intent. */
    public static final String ARG_LONGITUDE = BusStopMapFragment.ARG_LONGITUDE;
    
    private static final String DIALOG_MAP_TYPE_CHOOSER = "mapTypeDialog";
    private static final String DIALOG_SERVICES_CHOOSER =
            "servicesChooserDialog";
    private static final String DIALOG_PROGRESS = "progressDialog";
    
    /**
     * {@inheritDoc}
     */
    @Override
    public void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.single_fragment_container);

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        
        // Only add the fragment if there was no previous instance of this
        // Activity, otherwise this fragment will appear multiple times.
        if(savedInstanceState == null) {
            final Intent intent = getIntent();
            BusStopMapFragment f;
            
            if(Intent.ACTION_SEARCH.equals(intent.getAction())) {
                f = BusStopMapFragment.newInstanceWithSearch(
                        intent.getStringExtra(SearchManager.QUERY));
            } else if(intent.hasExtra(ARG_STOPCODE)) {
                f = BusStopMapFragment.newInstance(
                        intent.getStringExtra(ARG_STOPCODE));
            } else if(intent.hasExtra(ARG_LATITUDE) &&
                    intent.hasExtra(ARG_LONGITUDE)) {
                f = BusStopMapFragment.newInstance(
                        intent.getDoubleExtra(ARG_LATITUDE, 0),
                        intent.getDoubleExtra(ARG_LONGITUDE, 0));
            } else {
                f = new BusStopMapFragment();
            }
            
            getSupportFragmentManager().beginTransaction()
                    .add(R.id.fragmentContainer, f)
                    .commit();
        }
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public void onNewIntent(final Intent newIntent) {
        super.onNewIntent(newIntent);
        
        final BusStopMapFragment f = (BusStopMapFragment)
                getSupportFragmentManager()
                .findFragmentById(R.id.fragmentContainer);
        if(f == null) {
            return;
        }
        
        if(Intent.ACTION_SEARCH.equals(newIntent.getAction())) {
            f.onSearch(newIntent.getStringExtra(SearchManager.QUERY));
        } else if(newIntent.hasExtra(ARG_STOPCODE)) {
            f.moveCameraToBusStop(newIntent.getStringExtra(ARG_STOPCODE));
        } else if(newIntent.hasExtra(ARG_LATITUDE) &&
                newIntent.hasExtra(ARG_LONGITUDE)) {
            f.moveCameraToLocation(
                    new LatLng(newIntent.getDoubleExtra(ARG_LATITUDE, 0),
                        newIntent.getDoubleExtra(ARG_LONGITUDE, 0)),
                    BusStopMapFragment.DEFAULT_SEARCH_ZOOM, false);
        }
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        switch(item.getItemId()) {
            case android.R.id.home:
                NavigationUtils
                        .navigateUpOnActivityWithMultipleEntryPoints(this);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public void onBackPressed() {
        final BusStopMapFragment f = (BusStopMapFragment)
                getSupportFragmentManager()
                .findFragmentById(R.id.fragmentContainer);
        
        if(!f.onBackPressed()) {
            super.onBackPressed();
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void onShowMapTypeSelection() {
        new MapTypeChooserDialogFragment()
                .show(getSupportFragmentManager(), DIALOG_MAP_TYPE_CHOOSER);
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public void onShowServicesChooser(final String[] services,
                final String[] selectedServices, final String title) {
        ServicesChooserDialogFragment
                .newInstance(services, selectedServices, title)
                .show(getSupportFragmentManager(), DIALOG_SERVICES_CHOOSER);
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public void onShowSearchProgress(final String message) {
        IndeterminateProgressDialogFragment.newInstance(message)
                .show(getSupportFragmentManager(), DIALOG_PROGRESS);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void onDismissSearchProgress() {
        final IndeterminateProgressDialogFragment progressDialog =
                (IndeterminateProgressDialogFragment)
                        getSupportFragmentManager()
                                .findFragmentByTag(DIALOG_PROGRESS);
        if (progressDialog != null) {
            progressDialog.dismiss();
        }
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public void onShowBusStopDetails(final String stopCode) {
        final Intent intent = new Intent(this, BusStopDetailsActivity.class);
        intent.putExtra(BusStopDetailsActivity.ARG_STOPCODE, stopCode);
        startActivity(intent);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void onMapTypeChosen(final int mapType) {
        try {
            final MapTypeChooserDialogFragment.Callbacks child =
                    (MapTypeChooserDialogFragment.Callbacks)
                            getSupportFragmentManager()
                                    .findFragmentById(R.id.fragmentContainer);
            if (child != null) {
                child.onMapTypeChosen(mapType);
            }
        } catch (ClassCastException e) {
            // Unable to pass the callback on. Silently fail.
        }
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public void onServicesChosen(final String[] chosenServices) {
        try {
            final ServicesChooserDialogFragment.Callbacks child =
                    (ServicesChooserDialogFragment.Callbacks)
                            getSupportFragmentManager()
                                    .findFragmentById(R.id.fragmentContainer);
            if (child != null) {
                child.onServicesChosen(chosenServices);
            }
        } catch (ClassCastException e) {
            // Unable to pass the callback on. Silently fail.
        }
    }

    @Override
    public void onProgressCancel() {
        try {
            final IndeterminateProgressDialogFragment.Callbacks child =
                    (IndeterminateProgressDialogFragment.Callbacks)
                            getSupportFragmentManager()
                                    .findFragmentById(R.id.fragmentContainer);
            if (child != null) {
                child.onProgressCancel();
            }
        } catch (ClassCastException e) {
            // Unable to pass the callback on. Silently fail.
        }
    }
}

File: app/src/main/kotlin/uk/org/rivernile/android/bustracker/ui/busstopmap/BusStopMapActivity.kt
/*
 * Copyright (C) 2018 - 2025 Niall 'Rivernile' Scott
 *
 * This software is provided 'as-is', without any express or implied
 * warranty.  In no event will the authors or contributors be held liable for
 * any damages arising from the use of this software.
 *
 * The aforementioned copyright holder(s) hereby grant you a
 * non-transferrable right to use this software for any purpose (including
 * commercial applications), and to modify it and redistribute it, subject to
 * the following conditions:
 *
 *  1. This notice may not be removed or altered from any file it appears in.
 *
 *  2. Any modifications made to this software, except those defined in
 *     clause 3 of this agreement, must be released under this license, and
 *     the source code of any modifications must be made available on a
 *     publically accessible (and locateable) website, or sent to the
 *     original author of this software.
 *
 *  3. Software modifications that do not alter the functionality of the
 *     software but are simply adaptations to a specific environment are
 *     exempt from clause 2.
 *
 */

package uk.org.rivernile.android.bustracker.ui.busstopmap

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.ViewGroupCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updateLayoutParams
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.commit
import dagger.hilt.android.AndroidEntryPoint
import uk.org.rivernile.android.bustracker.ui.bustimes.DisplayStopDataActivity
import uk.org.rivernile.android.bustracker.ui.HasScrollableContent
import uk.org.rivernile.edinburghbustracker.android.R
import uk.org.rivernile.edinburghbustracker.android.databinding.ActivityBusStopMapBinding

/**
 * This [android.app.Activity] displays the stop map to the user. This is usually hosted by
 * [uk.org.rivernile.android.bustracker.ui.main.MainActivity]. However, this is used when another
 * part of the app, or an external app, wants to deeplink in to the map.
 *
 * This [android.app.Activity] merely hosts the [BusStopMapFragment]. That is where the map display
 * logic is performed.
 *
 * @author Niall Scott
 */
@AndroidEntryPoint
class BusStopMapActivity : AppCompatActivity(), BusStopMapFragment.Callbacks {

    companion object {

        const val EXTRA_STOP_CODE = "stopCode"
        const val EXTRA_LATITUDE = "latitude"
        const val EXTRA_LONGITUDE = "longitude"
    }

    private lateinit var viewBinding: ActivityBusStopMapBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        enableEdgeToEdge()
        viewBinding = ActivityBusStopMapBinding.inflate(layoutInflater)
        setContentView(viewBinding.root)
        ViewGroupCompat.installCompatInsetsDispatch(viewBinding.root)

        setSupportActionBar(viewBinding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        ViewCompat.setOnApplyWindowInsetsListener(viewBinding.toolbar) { view, windowInsets ->
            val insets = windowInsets.getInsets(
                WindowInsetsCompat.Type.systemBars() + WindowInsetsCompat.Type.displayCutout()
            )

            view.updateLayoutParams<ViewGroup.MarginLayoutParams> {
                leftMargin = insets.left
                rightMargin = insets.right
            }

            WindowInsetsCompat.CONSUMED
        }

        supportFragmentManager.registerFragmentLifecycleCallbacks(fragmentLifecycleCallbacks, true)

        if (savedInstanceState == null) {
            when {
                intent.hasExtra(EXTRA_STOP_CODE) ->
                    BusStopMapFragment.newInstance(intent.getStringExtra(EXTRA_STOP_CODE))
                intent.hasExtra(EXTRA_LATITUDE) && intent.hasExtra(EXTRA_LONGITUDE) -> {
                    val location = UiLatLon(
                        intent.getDoubleExtra(EXTRA_LATITUDE, 0.0),
                        intent.getDoubleExtra(EXTRA_LONGITUDE, 0.0)
                    )
                    BusStopMapFragment.newInstance(location)
                }
                else -> BusStopMapFragment.newInstance()
            }.let { fragment ->
                supportFragmentManager.commit {
                    add(R.id.fragmentContainer, fragment)
                }
            }
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)

        (currentFragment as? BusStopMapFragment)?.apply {
            when {
                intent.hasExtra(EXTRA_STOP_CODE) -> {
                    intent.getStringExtra(EXTRA_STOP_CODE)?.ifBlank { null }?.let {
                        onNewStopCode(it)
                        setIntent(intent)
                    }
                }
                intent.hasExtra(EXTRA_LATITUDE) && intent.hasExtra(EXTRA_LONGITUDE) -> {
                    val location = UiLatLon(
                        intent.getDoubleExtra(EXTRA_LATITUDE, 0.0),
                        intent.getDoubleExtra(EXTRA_LONGITUDE, 0.0)
                    )
                    onRequestCameraLocation(location)
                    setIntent(intent)
                }
            }
        }
    }

    override fun onShowBusTimes(stopCode: String) {
        Intent(this, DisplayStopDataActivity::class.java)
            .putExtra(DisplayStopDataActivity.EXTRA_STOP_CODE, stopCode)
            .let(this::startActivity)
    }

    /**
     * The current [androidx.fragment.app.Fragment] in the container.
     */
    private val currentFragment get() =
        supportFragmentManager.findFragmentById(R.id.fragmentContainer)

    private val fragmentLifecycleCallbacks = object : FragmentManager.FragmentLifecycleCallbacks() {
        override fun onFragmentResumed(fm: FragmentManager, f: Fragment) {
            viewBinding.appBarLayout.liftOnScrollTargetViewId =
                (f as? HasScrollableContent)?.scrollableContentIdRes ?: View.NO_ID
        }
    }
}
